Add a blog post for v1.10 (#3346)
FerretDB/FerretDB: website/blog/2023-09-13-ferretdb-v1-10-production-ready-sqlite.md

---
slug: ferretdb-v1-10-production-ready-sqlite
title: FerretDB v1.10.0 – Production-ready SQLite support and more
authors: [aleksi]
description: >
  We are happy to announce the full support for the SQLite backend and discuss our new architecture.
image: /img/blog/post-cover-image.jpg
tags: [release]
unlisted: true
---

<!-- FIXME banner image -->

We are happy to announce the full support for the SQLite backend and discuss our new architecture.

<!--truncate-->

We just released the new version of FerretDB – v1.10.0, which is [available in the usual places](https://docs.ferretdb.io/quickstart-guide/).
With it, the SQLite backend support is officially out of beta, on par with our PostgreSQL backend, and fully supported.
In the past two weeks, we added support for many missing commands and features,
including aggregation pipelines, indexes, query explains, and collection renaming.
Please refer to the [full release notes](https://github.com/FerretDB/FerretDB/releases/tag/v1.10.0) for details.

> There are still opportunities for improvement, such as support for more [query pushdowns](https://docs.ferretdb.io/pushdown/),

Reviewer: ```suggestion

There are still opportunities for improvement, such as support for more [query pushdowns](https://docs.ferretdb.io/pushdown/),
```

but they will not break the compatibility.
You can start to use the SQLite backend in production now!
It is ideal if you need a MongoDB-compatible database that does need high performance but could work in environments

> [where MongoDB can't](https://github.com/FerretDB/FerretDB/discussions/1266) and without a separate process.

Reviewer: I don't fully understand this sentence. It may need a bit of rewrite.

it supported a single PostgreSQL backend.
The simplified version of the architecture looked like this:

FIXME image

The protocol handling module was (and still is) responsible for handling client connections, MongoDB wire protocol,
and BSON documents encoding/decoding.
The commands handling module was responsible for handling protocol commands such as `find`, `update`, and `delete`.

Over time, we understood that there is some common code between command handlers.
We extracted a part responsible for accessing PostgreSQL and running queries into the backend package:

FIXME image

When we added another backend, we could not figure out if there was a common interface for different storages as they were so different.
So we decided to make another commands handling module that was similar in things like command parameters extraction
but very different in the way it interacted with storage.
Similar parts were extracted into a set of common packages:

FIXME image

Over time the common part grew, but we still had very different backend packages.
When we decided to add SQLite support, we wanted to revisit our previous decisions –
could we design a common storage interface after all?

> We did not expect many people to use SQLite with a lot of data, so performance could have a bit lower priority,

Reviewer: ```suggestion

We did not expect many people to use SQLite with a lot of data, so performance could have a bit lower priority,
```

and the interface could be relatively simple.
At the same time, we had a lot of practical knowledge of how FerretDB is used with PostgreSQL backend,
what is important and what is not, and what problems the current backend has and how to avoid them.
So we settled on this architecture:

FIXME image

> The common backend interface is relatively small to make it easy to add new backends.

Reviewer: ```suggestion
The common backend interface is relatively small and makes it easy to add new backends.
```

For example, there is a single method that backends should implement to insert given documents.
The code for processing ordered and unordered insertions is written once in the command handler;
there is no need to repeat that logic for every backend.
Of course, if there is some storage that could perform this operation more efficiently,
there could be a separate optional method that could be implemented.
Then the command handler could check if that method is implemented by the current backend and use it.

That architecture also allows us to add optional functionality to all backends at once.

> For example, there is already a _very early_ prototype of the OpLog functionality.

Reviewer: ```suggestion
For example, there is already a _very early_ prototype of the `OpLog` functionality.
```
